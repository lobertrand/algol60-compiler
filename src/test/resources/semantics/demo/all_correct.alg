begin

    begin
        integer a, b;
        ininteger(0, a);
        ininteger(0, b);

        begin
            comment Switch declaration;
            switch state := first, second;
            boolean chooseFirst;

            first:;
                outstring(1, "first operation");
                goto switchEnd;

            second:;
                outstring(1, "second operation");
                goto switchEnd;

            chooseFirst := b > 5 \/ a > 3 /\ a+b < 40;
            comment If-expression;
            goto state[if chooseFirst then 1 else 2]
        end;

        switchEnd:
    end;

    begin
        comment Array as parameter of procedure;
        integer procedure max(arr, a, b);
        value arr, a, b; integer array arr; integer a, b;
        begin
            integer i, result;
            result := arr[a];
            comment For-statement;
            for i := a+1 step 1 until b do
            begin
                comment If-statement;
                if arr[i] > result then
                    result := arr[i]
            end;
            max := result
        end;

        comment Array declaration and assignment;
        integer array a[1:3];
        a[1] := 6;
        a[2] := 8;
        a[3] := 3;

        comment Procedure call;
        outinteger(1, max(a, 1, 3))
    end;

    begin
        integer procedure getMessageNumber(grade);
        value grade; integer grade;
        begin
            comment If-statement exhaustiveness check;
            if grade < 8 then
                getMessageNumber := 1
            else if grade < 14 then
                getMessageNumber := 2
            else
                getMessageNumber := 3
        end;

        integer a;
        a := 1;
        begin
            string array messages[a:3];
            string msg;

            messages[1] := "bad";
            messages[2] := "okay";
            messages[3] := "great";

            msg := messages[getMessageNumber(12)];
        end
    end

end
