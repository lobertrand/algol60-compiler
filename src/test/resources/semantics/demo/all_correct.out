Algol60> Lexical analysis successful
Algol60> Syntactic analysis successful
┌───────────────────────────────────────────────┐
│ level: 0, tableNumber: 0                      │
├───────────────────────────────────────────────┤
│ Procedure: void outstring(integer, string)    │
│ Procedure: void outboolean(integer, boolean)  │
│ Procedure: void outinteger(integer, integer)  │
│ Procedure: void outreal(integer, real)        │
│ Procedure: void inreal(integer, real)         │
│ Procedure: void ininteger(integer, integer)   │
│ Procedure: integer entier(real)               │
└───────────────────────────────────────────────┘
    ┌──────────────────────────┐
    │ level: 1, tableNumber: 1 │
    ├──────────────────────────┤
    └──────────────────────────┘
        ┌──────────────────────────┐
        │ level: 2, tableNumber: 2 │
        ├──────────────────────────┤
        │ Variable: integer a      │
        │ Variable: integer b      │
        │ Label: switchEnd         │
        └──────────────────────────┘
            ┌────────────────────────────────┐
            │ level: 3, tableNumber: 3       │
            ├────────────────────────────────┤
            │ Switch: state [first, second]  │
            │ Variable: boolean chooseFirst  │
            │ Label: first                   │
            │ Label: second                  │
            └────────────────────────────────┘
        ┌──────────────────────────────────────────────────────────┐
        │ level: 2, tableNumber: 4                                 │
        ├──────────────────────────────────────────────────────────┤
        │ Procedure: integer max(integer array, integer, integer)  │
        │ Array: integer array a[1:3]                              │
        └──────────────────────────────────────────────────────────┘
            ┌──────────────────────────────────────────┐
            │ level: 3, tableNumber: 5                 │
            ├──────────────────────────────────────────┤
            │ Array: integer array arr[] (value param) │
            │ Variable: integer a (value param)        │
            │ Variable: integer b (value param)        │
            │ Variable: integer max (proc. result)     │
            │ Variable: integer i                      │
            │ Variable: integer result                 │
            └──────────────────────────────────────────┘
                ┌──────────────────────────┐
                │ level: 4, tableNumber: 6 │
                ├──────────────────────────┤
                └──────────────────────────┘
        ┌───────────────────────────────────────────────┐
        │ level: 2, tableNumber: 7                      │
        ├───────────────────────────────────────────────┤
        │ Procedure: integer getMessageNumber(integer)  │
        │ Variable: integer a                           │
        └───────────────────────────────────────────────┘
            ┌───────────────────────────────────────────────────┐
            │ level: 3, tableNumber: 8                          │
            ├───────────────────────────────────────────────────┤
            │ Variable: integer grade (value param)             │
            │ Variable: integer getMessageNumber (proc. result) │
            └───────────────────────────────────────────────────┘
            ┌────────────────────────────────────┐
            │ level: 3, tableNumber: 9           │
            ├────────────────────────────────────┤
            │ Array: string array messages[?:3]  │
            │ Variable: string msg               │
            └────────────────────────────────────┘

Algol60> Semantic analysis successful
